test/TokenPaymaster.test.js

```diff
@@ -0,0 +1,228 @@
+/* global contract artifacts before it */
+
+const TokenPaymaster = artifacts.require('TokenPaymaster.sol')
+const TestUniswap = artifacts.require('TestUniswap.sol')
+const TestToken = artifacts.require('TestToken.sol')
+const RelayHub = artifacts.require('RelayHub.sol')
+const TestProxy = artifacts.require('TestProxy')
+const { getEip712Signature } = require('../src/js/relayclient/utils')
+
+const RelayRequest = require('../src/js/relayclient/EIP712/RelayRequest')
+
+async function retcode (func) {
+  const ret = await func
+  const code = ret[0].toNumber()
+  const msg = code === 0 ? '' : ret[1] ? Buffer.from(ret[1].slice(2), 'hex').toString() : null
+  return {
+    code,
+    msg
+  }
+}
+
+function getLogs (receipt) {
+  let logs = receipt.logs || receipt
+  // retfirst = x => x[0]
+  if (!Array.isArray(logs)) {
+    logs = [logs]
+    // let retfirst = x => x
+  }
+  return (logs.map(log => ({
+    event: log.event,
+    ...Object.entries(log.args)
+      .filter(e => !/^[0-9_]/.test(e[0]))
+      .reduce((map, [key, val]) => ({
+        ...map,
+        [key]: val.toString()
+      }), {})
+  })))
+}
+
+contract('TokenPaymaster', ([from, relay, relayOwner]) => {
+  let paymaster, uniswap, token, recipient, hub
+  let sharedRelayRequestData
+
+  before(async () => {
+    // exchange rate 2 tokens per eth.
+    uniswap = await TestUniswap.new(2, 1, { value: 5e18 })
+    hub = await RelayHub.new(16)
+    token = await TestToken.at(await uniswap.tokenAddress())
+    paymaster = await TokenPaymaster.new(uniswap.address)
+    await paymaster.setRelayHub(hub.address)
+
+    // put some tokens in paymaster so it can calculate postRelayedCall gas usage:
+    await token.mint(1e18.toString())
+    await token.transfer(paymaster.address, 1e18.toString())
+    await paymaster.calculatePostGas()
+    console.log('withpre=', (await paymaster.gasUsedByPostWithPreCharge()).toString(), 'withoutpre=',
+      (await paymaster.gasUsedByPostWithoutPreCharge()).toString())
+
+    recipient = await TestProxy.new()
+    await recipient.setRelayHub(hub.address)
+
+    // approve uniswap to take our tokens.
+    await token.approve(uniswap.address, -1)
+
+    sharedRelayRequestData = {
+      senderAddress: from,
+      encodedFunction: recipient.contract.methods.test().encodeABI(),
+      senderNonce: '1',
+      target: recipient.address,
+      pctRelayFee: '1',
+      baseRelayFee: '0',
+      gasPrice: await web3.eth.getGasPrice(),
+      gasLimit: 1e6.toString(),
+      relayAddress: from,
+      paymaster: paymaster.address
+    }
+  })
+
+  context('#TestUniswap', () => {
```

Comment:
I am concerned these tests will interfere with tests on the paymaster itself. Also, they bloat the test case of the production component with a sanity test of the testing component. Can these be moved to a separate file?

Reply:
separated